src/Modules/Visualization/ShowField.cc

```diff
@@ -448,754 +538,293 @@ void GeometryBuilder::renderFacesLinear(
   int numAttributes = 3; //intially 3 because we will atleast be rendering verticies (vec3's)
 
   mesh->size(numFaces);
-  if (numFaces == 0)
-    return;
+  if (numFaces == 0) return;
 
-  bool withNormals = state.get(RenderState::USE_NORMALS);
-  if (withNormals)
+  bool useNormals = state.get(RenderState::USE_NORMALS);
+  bool useFaceNormals = state.get(RenderState::USE_FACE_NORMALS) && mesh->has_normals();
+  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  if (useNormals)
   {
     numAttributes += 3;
     mesh->synchronize(Mesh::NORMALS_E);
   }
 
-  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  bool useColorMap = (fld->basis_order() >= 0 && state.get(RenderState::USE_COLORMAP));
   ColorScheme colorScheme = ColorScheme::COLOR_UNIFORM;
-  std::vector<double> svals;
-  std::vector<Vector> vvals;
-  std::vector<Tensor> tvals;
-  std::vector<ColorRGB> face_colors;
 
-  if (fld->basis_order() < 0 || state.get(RenderState::USE_DEFAULT_COLOR))
-  {
-    colorScheme = ColorScheme::COLOR_UNIFORM;
-  }
-  else if (state.get(RenderState::USE_COLORMAP))
+  auto realColorMap = colorMap.get();
+  if(!realColorMap) realColorMap = StandardColorMapFactory::create();
+
+  auto map = StandardColorMapFactory::create(realColorMap->getColorMapName(),
+    realColorMap->getColorMapResolution(), realColorMap->getColorMapShift(),
+    realColorMap->getColorMapInvert());
+
+  auto coordinateMap = StandardColorMapFactory::create("Grayscale", 256, 0, false,
+    realColorMap->getColorMapRescaleScale(), realColorMap->getColorMapRescaleShift());
+
+  if (useColorMap)
   {
-    numAttributes += 4;
+    numAttributes += 2;
     colorScheme = ColorScheme::COLOR_MAP;
   }
-  else // if (fld->basis_order() >= 0)
-  {
-    numAttributes += 4;
-    colorScheme = ColorScheme::COLOR_IN_SITU;
-  }
 
   Point idpt;
   VMesh::Face::iterator fiter, fiterEnd;
   VMesh::Node::array_type nodes;
-  mesh->begin(fiter);
-  mesh->end(fiterEnd);
   mesh->get_nodes(nodes, *fiter);
   mesh->get_point(idpt, nodes[0]);
+  mesh->begin(fiter);
+  mesh->end(fiterEnd);
 
-  // Three 32 bit ints for each triangle to index into the VBO (triangles = verticies - 2)
-  size_t iboSize = static_cast<size_t>(mesh->num_faces() * sizeof(uint32_t) * (nodes.size() - 2) * 3);
-  size_t vboSize = static_cast<size_t>(mesh->num_faces() * sizeof(float) * nodes.size() * numAttributes);
-
-  // Construct VBO and IBO that will be used to render the faces. Once again, IBOs are not strictly
-  // needed. But, we may be able to optimize this code somewhat.
-  // todo Switch to unique_ptrs and move semantics.
-  std::shared_ptr<spire::VarBuffer> iboBufferSPtr(new spire::VarBuffer(iboSize));
-  std::shared_ptr<spire::VarBuffer> vboBufferSPtr(new spire::VarBuffer(vboSize));
+  int numNodes = nodes.size();
+  bool useQuads = (numNodes == 4);
```

Comment:
Is all the quad logic truly for only a 4-node mesh? Would it make much difference if tris were  used for such a small case?